x-pack/plugins/apm/server/routes/alerts/rule_types/error_count/register_error_count_rule_type.test.ts

```diff
@@ -126,11 +130,106 @@ describe('Error count alert', () => {
+      expect(services.alertFactory.create).toHaveBeenCalledWith(instanceName)
+    );
+
+    expect(scheduleActions).toHaveBeenCalledTimes(3);
+
+    expect(scheduleActions).toHaveBeenCalledWith('threshold_met', {
+      serviceName: 'foo',
+      environment: 'env-foo',
+      threshold: 2,
+      triggerValue: 5,
+      reason:
+        'Error count is 5 in the last 5 mins for foo, env-foo. Alert when > 2.',
+      interval: '5 mins',
+      viewInAppUrl:
+        'http://localhost:5601/eyr/app/apm/services/foo/errors?environment=env-foo',
+    });
+    expect(scheduleActions).toHaveBeenCalledWith('threshold_met', {
+      serviceName: 'foo',
+      environment: 'env-foo-2',
+      threshold: 2,
+      triggerValue: 4,
+      reason:
+        'Error count is 4 in the last 5 mins for foo, env-foo-2. Alert when > 2.',
+      interval: '5 mins',
+      viewInAppUrl:
+        'http://localhost:5601/eyr/app/apm/services/foo/errors?environment=env-foo-2',
+    });
+    expect(scheduleActions).toHaveBeenCalledWith('threshold_met', {
+      serviceName: 'bar',
+      environment: 'env-bar',
+      reason:
+        'Error count is 3 in the last 5 mins for bar, env-bar. Alert when > 2.',
+      threshold: 2,
+      triggerValue: 3,
+      interval: '5 mins',
+      viewInAppUrl:
+        'http://localhost:5601/eyr/app/apm/services/bar/errors?environment=env-bar',
+    });
+  });
+
+  it('sends alert when rule is configured with group by on transaction.name', async () => {
+    const { services, dependencies, executor, scheduleActions } =
+      createRuleTypeMocks();
+
+    registerErrorCountRuleType(dependencies);
+
+    const params = {
+      threshold: 2,
+      windowSize: 5,
+      windowUnit: 'm',
+      groupBy: ['service.name', 'service.environment', 'transaction.name'],
+    };
+
+    services.scopedClusterClient.asCurrentUser.search.mockResponse({
+      hits: {
+        hits: [],
+        total: {
+          relation: 'eq',
+          value: 2,
+        },
+      },
+      aggregations: {
+        error_counts: {
+          buckets: [
+            {
+              key: ['foo', 'env-foo', 'TRANSACTION_NAME_NOT_DEFINED'],
+              doc_count: 5,
+            },
+            {
+              key: ['foo', 'env-foo-2', 'tx-name-foo-2'],
+              doc_count: 4,
+            },
+            {
+              key: ['bar', 'env-bar', 'tx-name-bar'],
+              doc_count: 3,
+            },
+            {
+              key: ['bar', 'env-bar-2', 'tx-name-bar-2'],
+              doc_count: 1,
+            },
+          ],
+        },
+      },
+      took: 0,
+      timed_out: false,
+      _shards: {
+        failed: 0,
+        skipped: 0,
+        successful: 1,
+        total: 1,
+      },
+    });
+
     await executor({ params });
     [
-      'apm.error_rate_foo_env-foo',
-      'apm.error_rate_foo_env-foo-2',
-      'apm.error_rate_bar_env-bar',
+      'foo_env-foo_TRANSACTION_NAME_NOT_DEFINED',
```

Comment:
This should just be:

```suggestion
      'foo_env-foo,
```

But perhaps this point to another issue: perhaps what we actually want to pass in is:
```json
{
  `service.name`: `foo`,
  `service.environment`: `foo_env`
}
```

This way we know which fields the values belong to. And we can format them accordingly (eg formatting environment with `getEnvironmentLabel` from x-pack/plugins/apm/common/rules/apm_rule_types.ts)

Reply:
This is only for alert ID (not used anywhere user facing). Not sure why we should format it here?

Reply:
Either way, `TRANSACTION_NAME_NOT_DEFINED` should not be a thing.

Reply:
I formatted the reason message as per your suggestion. Previously I was not adding undefined fields in the reason message.

I saw `TRANSACTION_NAME_NOT_DEFINED` for missing transaction name bucket on CCS.